### ISSUE TITLE
Iterator::current() assumed not nullable

### ISSUE THREAD

**Issue by gmazzap** [vimeo/psalm#4141]
It seems that Psalm assumes `Iterator::current()` to always return the type assigned to the iterator via template annotations, but the reality is that `Iterator::current()` is always nullable because it returns null when the iterator is empty or consumed.

Consider the following code:

```php
/**
 * @return Iterator<string>
 */
function buildIterator(int $size): Iterator {
 
    $strings = [];
    for ($i = 0;  $i < $size; $i++) {
       $strings[] = random_bytes(1); 
    }
 
    return new ArrayIterator($strings);
}

$current = buildIterator(0)->current();
if (is_string($current)) {
    echo "First item is $current";
}
```

(https://psalm.dev/r/7db06f8ddd)

The condition `is_string($current)` is valid, because when the iterator is empty (and that's the case here) `$current` will be null (see https://3v4l.org/2job5).

Another example: https://psalm.dev/r/324b008117 this time false negative. The code will throw an exception if the iterator is empty (see https://3v4l.org/kDMSE) but for Psalm everything is fine.

---

In short, **for an iterator of type `Iterator<T>`, the method `Iterator::current()` has a return value of `T|null` and not `T`**.

Unless Psalm is able to infer that the iterator is not empty, e.g. in the following code:

```php
$iterator= buildIterator(2);
if ($iterator->valid()) {
   // In this case the condition is redundant because iterator is surely not empty at this point
   // so $current is surely a string according to the doc bloc
   if (is_string($current)) {
   }
}
```

Please note that this is not only valid for empty iterators, but also for non-empty iterators that have been consumed. For example, see here: https://3v4l.org/tmvX3 how for the same iterator instance `Iterator::current()` is either `string` or `null` depending on the iterator being valid or not.

**Comment by muglug** [comment]
Mind creating a PR on `CoreGenericClasses.phpstub`?

**Comment by gmazzap** [comment]
As you can see I've created #4146 that fixes stubs.

But I think this will disrupt a lot of codebases.

IMO it would be better to make use of `@psalm-assert-if-true` / `@psalm-assert-if-false` on the `Iterator::valid()` to assert that the `Iterator::current()` will return `TValue` when `Iterator::valid()` returns `true` and will return `null`  when when `Iterator::valid()` returns `false`.

To introduce this assertion I think it is required to add a new template type `TCurrent` of type `TValue|null` so that `current()` could return `TCurrent`. But I'm not sure this is doable...

**Comment by gmazzap** [comment]
I said "will disrupt a lot of codebases" and I guess that the fact that Psalm's own files don't pass Psalm check with this commit is a sign of that.

**Comment by weirdan** [comment]
Wouldn't adding [@psalm-ignore-nullable-returns](https://psalm.dev/docs/annotating_code/supported_annotations/#psalm-ignore-nullable-return) fix that?

**Comment by gmazzap** [comment]
"Fix" as in "don't reported" yes. But then if you have a case where `current` returns `null` you get in trouble* (see https://3v4l.org/kDMSE).

So what you've to do is:

```php
if ($it->valid()) {
  /** @psalm-ignore-nullable-returns */
  return $it->current();
}
```

which surely works, but would not be better if Psalm would be able to figure it out without any additional annotation?

---

* when you receive an Iterator as a parameter there's no way to know it is empty or it was already consumed, so there's no way to know if `current` will return `null` or not. This means that if you ignore the possibility of a nullable return, you're ignoring a real issue of your code. Sure thing, if you instantiate a non-empty iterator right before calling `current` you're sure that `current()` will not return `null`, but even in that case you need to be very careful, see https://3v4l.org/1B1Xs

**Comment by muglug** [comment]
Fixed in f78fbbe

**Comment by muglug** [comment]
I followed @weirdan's recommendation and allowed Psalm to ignore the issue – e.g. in the `while ($it->valid())` example you gave:

```php
<?php

/**
 * @return Iterator<string>
 */
function buildIterator(int $size): Iterator {
    $values = [];
    for ($i = 0;  $i < $size; $i++) {
       $values[] = "Item $i\n"; 
    }
 
    return new ArrayIterator($values);
}

$it = buildIterator(2);

while($it->valid()) {
    echo strlen($it->current());
}
```

I appreciate that this is not exactly what you wanted, but I err on the side of not labelling good code as being buggy, when there isn't currently a way to work around it (adding some sort of typestate-based solution might work in the future, but the mechanics of that are still TBD).

**Comment by gmazzap** [comment]
@muglug The problem for me is things like this: https://psalm.dev/r/324b008117 when Psalm says "No issues" but the code is actually bugged (see https://3v4l.org/kDMSE).

This is not something I can fix with annotations of sort.

**Comment by muglug** [comment]
> `Iterator::current()` is at any effect nullable. When used in contexts where null is not an option, then I expect an error to be raised.

Yeah, but that's problematic for code that works today (e.g. with those `valid()` checks). I don't want people to have to write workarounds just to please Psalm.

**Comment by gmazzap** [comment]
I definetively understand that, and I was the first saying that to just declare the stub as nullable would **not** be a good solution.

IMO the point is not to "please Psalm", but to find issues in code.

Anyone that has a function that declares return type `string` but returns the result of `Iterator::current()` should not change that code to "please Psalm", but to fix a a _real_ code issue... one of those issues that are the reason why people install Psalm in first place.

As I already said, Psalm is able to understand when the `Iterator::current` will return null, that is when `Iterator::valid()` returns `false`.

By using a combination of `@psalm-assert-if-true` and  `@psalm-assert-if-false` on `@Iterator::valid()`, would be possible to conditionally assert the value of a template type `TCurrent` and use that as return type of `Iterator::current()`.

To make more clear what I mean, look at https://psalm.dev/r/3faef1610d See how `test()` function return type is correctly understood? That is thanks to the `@psalm-assert-if-true` I put on `valid()` method.

For a comparison, look at: https://psalm.dev/r/e44b7bb7ab There I removed the `@psalm-assert-if-true` from `valid()` and Psalm complains for `InvalidNullableReturnType` for `test()`.

For me the best option for iterators would be the introduction in stubs of `TCurrent` and assertions on `Iterator::valid()` that would assert `TCurrent` to be either `TValue` or null depending on `valid()` returning true or not. Just like my example https://psalm.dev/r/3faef1610d

I would create a PR with this, but to be honest I'm not sure how that could be done from a stub.
